tests/core/test_jose/test_jwe.py

```diff
@@ -238,6 +251,135 @@ def test_ecdh_es_raise(self):
+            "kty": "EC",
+            "crv": "P-256",
+            "x": "gI0GAILBdu7T53akrFmMyGcsF3n5dO7MmwNBHKW5SV0",
+            "y": "SLW_xSffzlPWrHEVI30DHM_4egVwt3NQqeUD7nMFpps",
+            "d": "0_NxaRPUMQoAJt50Gz8YiTr8gRTwyEaCumd-MToTmIo"
+        }
+        headers = {
+            "alg": "ECDH-1PU",
+            "enc": "A256GCM",
+            "apu": "QWxpY2U",
+            "apv": "Qm9i"
+        }
+
+        alg = JsonWebEncryption.ALG_REGISTRY['ECDH-1PU']
+
+        alice_static_key = alg.prepare_key(alice_static_key)
+        bob_static_key = alg.prepare_key(bob_static_key)
+        alice_ephemeral_key = alg.prepare_key(alice_ephemeral_key)
+
+        alice_static_pubkey = alice_static_key.get_op_key('wrapKey')
+        bob_static_pubkey = bob_static_key.get_op_key('wrapKey')
+        alice_ephemeral_pubkey = alice_ephemeral_key.get_op_key('wrapKey')
+
+        dk_at_alice = alg.deliver_at_sender(alice_static_key, alice_ephemeral_key, bob_static_pubkey, headers, 256, None)
+        self.assertEqual(urlsafe_b64encode(dk_at_alice), b'bK8Tcj0UhQrUtCzW3ek1v_0v_wCpunDeBcIDpeFyLKc')
+
+        dk_at_bob = alg.deliver_at_recipient(bob_static_key, alice_static_pubkey, alice_ephemeral_pubkey, headers, 256, None)
+        self.assertEqual(urlsafe_b64encode(dk_at_bob), b'bK8Tcj0UhQrUtCzW3ek1v_0v_wCpunDeBcIDpeFyLKc')
+
+    def test_ecdh_1pu_jwe(self):
+        jwe = JsonWebEncryption()
+        alice_key = {
+            "kty": "EC",
+            "crv": "P-256",
+            "x": "WKn-ZIGevcwGIyyrzFoZNBdaq9_TsqzGl96oc0CWuis",
+            "y": "y77t-RvAHRKTsSGdIYUfweuOvwrvDD-Q3Hv5J0fSKbE",
+            "d": "Hndv7ZZjs_ke8o9zXYo3iq-Yr8SewI5vrqd0pAvEPqg"
+        }
+        bob_key = {
+            "kty": "EC",
+            "crv": "P-256",
+            "x": "weNJy2HscCSM6AEDTDg04biOvhFhyyWvOHQfeF_PxMQ",
+            "y": "e8lnCO-AlStT-NJVX-crhB7QRYhiix03illJOVAOyck",
+            "d": "VEmDZpDXXK8p8N0Cndsxs924q6nS1RXFASRl6BfUqdw"
+        }
+        for alg in ["ECDH-1PU", "ECDH-1PU+A128KW", "ECDH-1PU+A192KW", "ECDH-1PU+A256KW"]:
+            protected = {'alg': alg, 'enc': 'A256GCM'}
+            data = jwe.serialize_compact(protected, b'hello', bob_key, sender_key=alice_key)
+            rv = jwe.deserialize_compact(data, bob_key, sender_key=alice_key)
+            self.assertEqual(rv['payload'], b'hello')
+
+    def test_ecdh_1pu_with_okp(self):
+        jwe = JsonWebEncryption()
+        alice_key = OKPKey.generate_key('X25519', is_private=True)
+        bob_key = OKPKey.generate_key('X25519', is_private=True)
+        for alg in ["ECDH-1PU", "ECDH-1PU+A128KW", "ECDH-1PU+A192KW", "ECDH-1PU+A256KW"]:
+            protected = {'alg': alg, 'enc': 'A256GCM'}
+            data = jwe.serialize_compact(protected, b'hello', bob_key, sender_key=alice_key)
+            rv = jwe.deserialize_compact(data, bob_key, sender_key=alice_key)
+            self.assertEqual(rv['payload'], b'hello')
+
+    def test_ecdh_1pu_raise(self):
```

Comment:
I believe we may need to add more tests (and possibly checks in implementation) here:
1) Make sure that ECDH-1PU+AXXXKW algorithms can work with CBC encryption algorithms only (A256CBC-HS512 etc.). Direct key mode can work with any algorithms
2) A test that raises an exception if valid keys are passed (private and public) but of different type (JWK key type field)
3) A test that raises an exception if the key types are valid, the format is valid, but the value of the points (x, y) doesn't belong to the same curve.